core/source.ts

```diff
@@ -321,15 +333,43 @@ export default class SiteSource implements Source {
 
   load(path: string, loader: Loader): Promise<Data> {
     try {
-      if (!this.#cache.has(path)) {
-        this.#cache.set(path, loader(path));
+      if (this.#cache.has(path)) {
+        return this.#cache.get(path)!;
       }
 
-      return this.#cache.get(path)!;
+      const promise = loader(path);
+      this.#cache.set(path, promise);
+
+      promise.then((data) => this.#updateDependencies(path, data));
+
+      return promise;
     } catch (cause) {
       throw new Exception("Couldn't load this file", { cause, path });
     }
   }
+
+  #updateDependencies(path: string, { dependencies }: Data) {
+    const toLocal = absolutePathToLocal(this.site.src());
+    const currentFile = toLocal(path);
+
+    for (const [_, dependants] of this.#dependants) {
+      dependants.delete(currentFile);
+    }
```

Comment:
First we clean up the current file from any dependency it had before. If it's still a dependency we're going to add it back anyway.